File: hasekll/Stack.hs
module Stack 
    (
     newStack,
     isEmpty,
     peek,
     pop,
     push,
     size,
    )
where

data Stack a = Stack [a]

newStack :: Stack a
newStack = Stack []

isEmpty :: Stack a -> Bool
isEmpty (Stack s) = null s

peek :: Stack a -> a
peek (Stack s) = head s

pop :: Stack a -> (a, Stack a)
pop (Stack s) = (head s, Stack $ tail s)

push :: a -> Stack a -> Stack a
push e (Stack s) = Stack $ e : s

size :: Stack a -> Int
size (Stack s) = length s

File: c/tautology/stack.c
#include "stack.h"
#include <stdlib.h>
#include <strings.h>

void *get_top(struct stack *stack)
{
	return (*(stack->st))[stack->top];
}

void *pop(struct stack *stack)
{
	return (*(stack->st))[(stack->top)--];
}

void push(struct stack *stack, void *elem)
{
	(*(stack->st))[++(stack->top)] = elem;
}

int is_empty(struct stack* stack)
{
	return stack->top == -1;
}

struct stack *new_stack(int count)
{
	struct stack *st = malloc(sizeof(struct stack));
	st->top = -1;
	st->st  = malloc(count * sizeof(void*));
	bzero(st->st, count * sizeof(void*));
	return st;
}
